### ISSUE TITLE
Roslyn Test Failures in VS 2015 RTM

### ISSUE THREAD

**Issue by daveaglick** [hazzik/DelegateDecompiler#54]
I was trying to reproduce a bug I was having only when building/running under VS 2015 RTM so I opened DelegateDecompiler to create a test and discovered a bunch of the existing tests also don't work in VS 2015. I'll assume these are due to Roslyn IL differences. See the screen shot below. I'm going to start trying to knock them out, but some of these may stretch the limits of my IL abilities.

![2015-07-23_10h37_57](https://cloud.githubusercontent.com/assets/1020407/8853054/e6120834-3126-11e5-8f06-bcc031fee282.png)

**Comment by daveaglick** [comment]
FYI - for the record, here's the IL in VS 2013 for `CanUseLambda`:

```
IL_0000: ldc.i4.1
IL_0001: ldc.i4.s 10
IL_0003: call System.Collections.Generic.IEnumerable`1[System.Int32] Range(Int32, Int32)
IL_0008: ldsfld System.Func`2[System.Int32,System.Boolean] CS$<>9__CachedAnonymousMethodDelegate3
IL_000d: brtrue.s IL_0022
IL_000f: ldnull
IL_0010: ldftn Boolean <CanUseLambda>b__1(Int32)
IL_0000: ldarg.0
IL_0001: ldc.i4.5
IL_0002: ceq
IL_0004: stloc.0
IL_0005: br.s IL_0007
IL_0007: ldloc.0
IL_0008: ret
IL_001b: stsfld System.Func`2[System.Int32,System.Boolean] CS$<>9__CachedAnonymousMethodDelegate3
IL_0020: br.s IL_0022
IL_0022: ldsfld System.Func`2[System.Int32,System.Boolean] CS$<>9__CachedAnonymousMethodDelegate3
IL_0027: call Boolean Any[Int32](System.Collections.Generic.IEnumerable`1[System.Int32], System.Func`2[System.Int32,System.Boolean])
IL_002c: stloc.0
IL_002d: br.s IL_002f
IL_002f: ldloc.0
IL_0030: ret
```

And here it is in Roslyn 1.0/VS 2015 RTM (prior to throwing):

```
IL_0000: ldc.i4.1
IL_0001: ldc.i4.s 10
IL_0003: call System.Collections.Generic.IEnumerable`1[System.Int32] Range(Int32, Int32)
IL_0008: ldsfld System.Func`2[System.Int32,System.Boolean] <>9__0_1
IL_000d: dup
IL_000e: brtrue.s IL_0027
IL_0010: pop
IL_0011: ldsfld <>c <>9
IL_0016: ldftn Boolean <CanUseLambda>b__0_1(Int32)
IL_0000: ldarg.1
IL_0001: ldc.i4.5
IL_0002: ceq
IL_0004: ret
IL_0021: dup
IL_0022: stsfld System.Func`2[System.Int32,System.Boolean] <>9__0_1
System.ArgumentException : Argument types do not match
```

**Comment by daveaglick** [comment]
I suspect the failed lambda tests have something to do with changes to how anonymous delgates are cached: http://stackoverflow.com/questions/30897647/delegate-caching-behavior-changes-in-roslyn

In debugging side-by-side, the first sign of major differences is in the `brtrue` handler at this condition:

```
if (memberExpression != null && IsCachedAnonymousMethodDelegate(memberExpression.Member as FieldInfo))
{
    state.Stack.Pop();
}
else
{
    state.Instruction = ConditionalBranch(state, val => Expression.NotEqual(val, Default(val.Type)));
    continue;
}
```

In VS 2013 it evaluates to true because the anonymous delegate with the `i == 5` condition was at the top of the stack. In VS 2015, we end up adding a `ConditionalBranch` instruction because of the test to set the lifted `Func<int, bool>` in the generated class if it's not already set.

I'm finding this helpful to see what's going on as far as what Roslyn is generating:
http://tryroslyn.azurewebsites.net/#K4Zwlgdg5gBAygTxAFwKYFsDcAoUlaIoYB0AwgPYA2lqAxsmORCMQOKoSoBOYtOe0eEjTpiAGUgBHfuEGER4qcQCiADwAOXVCHBMQOdcABGlXjFqUAhjpikYAb2wxnMQybMA3cmAAmMALIAFACUME4uji5RMMoQwOjcliaoxABKltCogQCMADQw2QAMwcQAghAIgWAwALwAfDDVNTUwAKzBOFEAvtg9QAAA=

**Comment by daveaglick** [comment]
At this point I'll wait and see if you have an easy fix or know exactly what's going on. I don't want to spend too much time diving in deep if it's obvious to you where the problem lies. If not, I don't mind continuing to dig.

**Comment by daveaglick** [comment]
Couldn't help myself... I think I've got the lambda issues resolved. I changed `IsCachedAnonymousMethodDelegate` to check for the field name starting with either `CS$<>9__` (old compiler with cached anonymous delegate) or `<>9__` (Roslyn compiler with lifted anonymous delegate). This should be safe given the pattern `P<N>C__SI` specified at http://stackoverflow.com/a/2509524/807064 still appears to hold true for Roslyn.

Curiosity has gotten the better of me, so I'm going to keep looking at some of the other issues for now.

**Comment by daveaglick** [comment]
The failures with enums is a little odd. It appears as though DelegateDecompiler and Roslyn are actually generating the same thing they did under the old compiler. It's the expected test value that's different - Roslyn apparently triggers a conversion of the enum value in the expression body. If you paste the following code into https://dotnetfiddle.net and then select either the old compiler or Roslyn, you'll see what I mean:

```
using System;
using System.Linq.Expressions;
					
public class Program
{
	public enum TestEnum
    {
        Foo,
        Bar
    }
	
	public static void Main()
	{
		Expression<Func<TestEnum, bool>> expected = x => TestEnum.Bar == x;
		Console.WriteLine(expected.Body.ToString());
	}
}
```

The old compiler outputs `(1 == Convert(x))`, which is also what DelegateDecompiler produces under either compiler. But Roslyn outputs `(Convert(Bar) == Convert(x))` for the expected body.

**Comment by hazzik** [comment]
Thanks, @daveaglick!

**Comment by hazzik** [comment]
Fixed in 0.17.0
